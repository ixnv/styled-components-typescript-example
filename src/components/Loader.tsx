import * as React from 'react'
import styled, {Keyframes, keyframes} from "styled-components";

type LoaderProps = {
    show: boolean
}

const Wrapper = styled.div`
    display: ${(props: LoaderProps) => props.show ? 'block' : 'none'}
`

const Loader = styled.div`
    position: relative;
    display: inline-block;
    height: 36px;
    width: 64px;

    & div {
        animation-timing-function: cubic-bezier(0, 1, 1, 0);
        background: #2a5885;
        border-radius: 50%;
        height: 11px;
        position: absolute;
        top: 18px;
        width: 11px;
    }
`

const piece1Animation = keyframes`
    0% {
        transform: scale(0);
    }
    100% {
        transform: scale(1);
    }
`

const piece2Animation = keyframes`
    0% {
        transform: translate(0, 0);
    }
    100% {
        transform: translate(19px, 0);
    }
`

const piece3Animation = keyframes`
    0% {
        transform: scale(1);
    }
    100% {
        transform: scale(0);
    }
`

type AnimationProps = {
    childIndex: number
    animation: Keyframes
    left: string
}

const Piece = styled(Loader)`
    &:nth-child(${(props: AnimationProps) => props.childIndex}) {
        animation: ${(props: AnimationProps) => props.animation} 0.6s infinite;
        left: ${(props: AnimationProps) => props.left};
        will-change: transform;
    }
`

const LoaderComponent: React.FC<LoaderProps> = ({show}: LoaderProps) => (
    <Wrapper show={show}>
        <Loader>
            <Piece childIndex={1} animation={piece1Animation} left='6px'/>
            <Piece childIndex={2} animation={piece2Animation} left='6px'/>
            <Piece childIndex={3} animation={piece2Animation} left='26px'/>
            <Piece childIndex={4} animation={piece3Animation} left='45px'/>
        </Loader>
    </Wrapper>
)

export default LoaderComponent;
