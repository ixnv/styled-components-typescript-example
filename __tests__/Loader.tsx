import * as React from "react";
import { shallow } from 'enzyme'
import Loader from "../src/components/Loader";

test('<Loader/> renders without crash', () => {
    const loader = shallow(<Loader show={true}/>)

    expect(loader).toMatchSnapshot()
})