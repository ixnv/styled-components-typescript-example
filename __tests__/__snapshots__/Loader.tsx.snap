// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader/> renders without crash 1`] = `
<styled.div
  show={true}
>
  <styled.div>
    <Styled(styled.div)
      animation={
        Keyframes {
          "id": "sc-keyframes-bBWfDF",
          "inject": [Function],
          "name": "bBWfDF",
          "rules": Array [
            "@-webkit-keyframes bBWfDF{0%{-webkit-transform:scale(0);-ms-transform:scale(0);transform:scale(0);}100%{-webkit-transform:scale(1);-ms-transform:scale(1);transform:scale(1);}}",
            "@keyframes bBWfDF{0%{-webkit-transform:scale(0);-ms-transform:scale(0);transform:scale(0);}100%{-webkit-transform:scale(1);-ms-transform:scale(1);transform:scale(1);}}",
          ],
          "toString": [Function],
        }
      }
      childIndex={1}
      left="6px"
    />
    <Styled(styled.div)
      animation={
        Keyframes {
          "id": "sc-keyframes-eEJBHY",
          "inject": [Function],
          "name": "eEJBHY",
          "rules": Array [
            "@-webkit-keyframes eEJBHY{0%{-webkit-transform:translate(0,0);-ms-transform:translate(0,0);transform:translate(0,0);}100%{-webkit-transform:translate(19px,0);-ms-transform:translate(19px,0);transform:translate(19px,0);}}",
            "@keyframes eEJBHY{0%{-webkit-transform:translate(0,0);-ms-transform:translate(0,0);transform:translate(0,0);}100%{-webkit-transform:translate(19px,0);-ms-transform:translate(19px,0);transform:translate(19px,0);}}",
          ],
          "toString": [Function],
        }
      }
      childIndex={2}
      left="6px"
    />
    <Styled(styled.div)
      animation={
        Keyframes {
          "id": "sc-keyframes-eEJBHY",
          "inject": [Function],
          "name": "eEJBHY",
          "rules": Array [
            "@-webkit-keyframes eEJBHY{0%{-webkit-transform:translate(0,0);-ms-transform:translate(0,0);transform:translate(0,0);}100%{-webkit-transform:translate(19px,0);-ms-transform:translate(19px,0);transform:translate(19px,0);}}",
            "@keyframes eEJBHY{0%{-webkit-transform:translate(0,0);-ms-transform:translate(0,0);transform:translate(0,0);}100%{-webkit-transform:translate(19px,0);-ms-transform:translate(19px,0);transform:translate(19px,0);}}",
          ],
          "toString": [Function],
        }
      }
      childIndex={3}
      left="26px"
    />
    <Styled(styled.div)
      animation={
        Keyframes {
          "id": "sc-keyframes-gjACFv",
          "inject": [Function],
          "name": "gjACFv",
          "rules": Array [
            "@-webkit-keyframes gjACFv{0%{-webkit-transform:scale(1);-ms-transform:scale(1);transform:scale(1);}100%{-webkit-transform:scale(0);-ms-transform:scale(0);transform:scale(0);}}",
            "@keyframes gjACFv{0%{-webkit-transform:scale(1);-ms-transform:scale(1);transform:scale(1);}100%{-webkit-transform:scale(0);-ms-transform:scale(0);transform:scale(0);}}",
          ],
          "toString": [Function],
        }
      }
      childIndex={4}
      left="45px"
    />
  </styled.div>
</styled.div>
`;
